### PR TITLE
fix(ast): format with good parens

### PR DESCRIPTION
**Pull request by affo**
Closes https://github.com/influxdata/flux/issues/1012

### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

**Comment by affo** [comment]
@juliusv You are right and thanks for pointing it out.

`1 - (2 - 3)` is short for `1 - 1 * (2 - 3)`.

My algorithm worked for the second one but not for the first one. It required custom logic, because that is the only case in which you need to check AST composition to put parenthesis.

File: internal/parser/parser_test.go [hunk 1]
```diff
@@ -2096,6 +2096,134 @@ a = 5.0
...
+								Right: &ast.IntegerLiteral{
+									BaseNode: base("1:10", "1:11"),
+									Value:    3,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
```
Comment (affo) [inline internal/parser/parser_test.go]: I added these tests in the parser while studying AST generation and I thought about leaving them there.

Surely, they can't hurt.

File: ast/format_test.go [hunk 2]
```diff
@@ -325,6 +328,29 @@ func TestFormat_Associativity(t *testing.T) {
...
+			shouldFail: true,
+		},
+		{
+			name:       "plus with unintended parens 2",
+			script:     `1 + (2 + 3)`,
+			shouldFail: true,
+		},
+		{
+			name:       "minus with unintended parens",
+			script:     `(1 - 2) - 3`,
```
Comment (jlapacik) [inline ast/format_test.go]: Could you add a test case for `1 - 2 - 3`?

File: internal/parser/parser_test.go (hunk 1 above)
Comment (jlapacik) [reply internal/parser/parser_test.go]: Definitely

File: ast/format.go [hunk 3]
```diff
@@ -122,7 +122,59 @@ var opPrecedence = map[int]int{
 	getIntForLOp(OrOperator):  7,
 }
 
+// About parenthesis:
+// As a golden rule, we need parenthesis if a child node has lower precedence (bigger value) then its parent node.
```
Comment (jlapacik) [inline ast/format.go]: nit: then -> than

File: ast/format.go [hunk 4]
```diff
@@ -122,7 +122,59 @@ var opPrecedence = map[int]int{
...
+	f.formatLeftChildWithParens(parent, child)
+}
+
+// formatLeftChildWithParens applies the golden rule for parenthesis.
+func (f *formatter) formatLeftChildWithParens(parent, child Node) {
+	pvp, pvc := getPrecedences(parent, child)
+	// If one of parent or child has not matched any case, then we shouldn't apply any parenthesis.
+	par := !(pvc == 0 || pvp == 0)
+	// We need a parenthesis if parent and child have matched, and if the child node
+	// has lower precedence (bigger value) then the current node.
```
Comment (jlapacik) [inline ast/format.go]: nit: then -> than

File: ast/format.go [hunk 5]
```diff
@@ -154,18 +179,37 @@ func (f *formatter) formatChildWithParens(parent, child Node) {
...
+// (C) - When there are parenthesis, instead, the parser recursively generates a AST for the expression contained
+// in the parenthesis, and makes it the right child.
+// So, when formatting:
+//  - if we encounter a child with lower precedence on the left, this means it requires parenthesis, because, for sure,
+//    the parser detected parenthesis to break (A);
+//  - if we encounter a child with higher or equal precedence on the left, it doesn't need parenthesis, because
+//    that was the natural parsing order of elements (see (B));
+//  - if we encounter a child with lower or equal precedence on the right, it requires parenthesis, otherwise, it
+//    would have been at root (see (C)).
+func needsPar(pvp, pvc int, isRight bool) bool {
```
Comment (jsternberg) [inline ast/format.go]: Can you rename this something that's more descriptive? Maybe just spell out the entire word.

**Pull request merged**